File: Python/input_ham.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-
from __future__ import print_function, division
import numpy as np
"""
fname: input file name
sw_hoplist: switch hopping order if True [nr,no,no] else [no,no,nr] default True
no is number of orbitals. nr is number of hopping matrices. 
sw_ndegen: import_hop only, select lead ndegen from ndegen.txt (True) or not (False) default False
"""
def import_hop(fname,sw_ndegen=False,sw_hoplist=True):
    rvec=np.loadtxt(fname+'/irvec.txt')
    nr=rvec[:,0].size
    tmp=np.array([complex(float(tp[0]),float(tp[1])) for tp in [f.strip(' ()\n').split(',') for f in open(fname+'/ham_r.txt','r')]])
    no=int(np.sqrt(tmp.size/nr))
    ham_r=(tmp.reshape(nr,no,no) if sw_hoplist else np.reshape(tmp,(nr,no*no)).T.reshape(no,no,nr))
    ndegen=(np.loadtxt(fname+'/ndegen.txt') if sw_ndegen else np.ones(nr))
    return(rvec,ndegen,ham_r,no,nr)

def import_out(fname,sw_hoplist=True):
    data=np.loadtxt(fname)
    con=(data[:,:3]==data[0,:3]).prod(axis=1).sum()
    no,nr =int(np.sqrt(con)),data[:,0].size//con
    rvec=data[:nr,:3]
    ham_r=(data[:,3]+1j*data[:,4]).reshape((nr,no,no) if sw_hoplist else (no,no,nr))
    ndegen=np.ones(nr)
    return(rvec,ndegen,ham_r,no,nr)

def import_hr(name,sw_hoplist=True):
    tmp=[f.split() for f in open('%s_hr.dat'%name,'r')]
    no, nr=int(tmp[1][0]), int(tmp[2][0])
    c2,tmp1=3,[]
    while not len(tmp1)==nr:
        tmp1.extend(tmp[c2])
        c2=c2+1
    ndegen=np.array([int(t) for t in tmp1])
    tmp1=[[float(t) for t in tp] for tp in tmp[c2:]]
    tmp=np.array([complex(tp[5],tp[6]) for tp in tmp1])
    rvec=np.array([tmp1[no*no*i][:3] for i in range(nr)])
    ham_r=(np.reshape(tmp,(nr,no,no)) if sw_hoplist
           else np.reshape(tmp,(nr,no*no)).T.reshape(no,no,nr))
    return(rvec,ndegen,ham_r,no,nr)

def import_Hopping(sw_hoplist=True,sw_axis=False):
    tmp=[f.split() for f in open('Hopping.dat','r')]
    axis=np.array([[float(tp) for tp in tpp] for tpp in tmp[1:4]])
    no,nr=int(tmp[4][0]),int(tmp[4][1])
    ndegen=np.array([1]*nr)
    tmp1=[[float(t) for t in tp] for tp in tmp[7+no:]]
    rvec=np.array([tmp1[no*no*i][:3] for i in range(nr)])
    tmp=np.array([complex(tp[8],tp[9]) for tp in tmp1])
    ham_r=(np.reshape(tmp,(nr,no,no)) if sw_hoplist
           else np.reshape(tmp,(nr,no*no)).T.reshape(no,no,nr))
    if sw_axis:
        return(rvec,ndegen,ham_r,no,nr,axis)
    else:
        return(rvec,ndegen,ham_r,no,nr)

File: Julia/input_ham.jl
#!/usr/bin/env julia
#=
Julia input Hamiltonian code
=#
function import_hop(fname,sw_ndegen)
    tmp=open(fname*"/irvec.txt","r") do fp
        readlines(fp)
    end
    nr=length(tmp)
    rvec=zeros(Float64,3,nr)
    for i in 1:nr
        rvec[:,i]=parse.(Float64,split(tmp[i]))
    end
    tmp=open(fname*"/ham_r.txt","r") do fp
        readlines(fp)
    end
    no=Int(sqrt(length(tmp)/nr))
    tmp2=[parse.(Float64,split(strip(l,['\n','(',')']),',')) for l in tmp]
    tmp=[t[1]+1im*t[2] for t in tmp2]
    ham_r=zeros(Complex,nr,no,no)
    for i in 1:no
        for j in 1:no
            for k in 1:nr
                ham_r[k,j,i]=tmp[i+no*(j-1)+no*no*(k-1)]
            end
        end
    end

    if sw_ndegen
        tmps=open(fname*"/ndegen.txt","r") do fp
            readlines(fp)
        end
        ndegen=parse.(Float64,tmp)
    else
        ndegen=ones(Float64,(1,nr))
    end 
    return(rvec,ndegen,ham_r,no,nr)
end

function import_out(fname)
    tmp=open(fname,"r") do fp
        readlines(fp)
    end
    tmp1=[parse.(Float64,split(l)) for l in tmp]
    r0=tmp1[1][1:3]
    con=0
    for rr in tmp1
        if rr[1:3]==r0
            con+=1
        end
    end
    no=Int(sqrt(con))
    nr=div(length(tmp1),con)
    rvec=zeros(Float64,3,nr)
    for i in 1:nr
        for j in 1:3
            rvec[j,i]=tmp1[i][j]
        end
    end
    tmp=[tp[4]+1im*tp[5] for tp in tmp1]
    ham_r=zeros(Complex,nr,no,no)
    for i in 1:no
        for j in 1:no
            for k in 1:nr
                ham_r[k,j,i]=tmp[k+nr*(j-1)+no*nr*(i-1)]
            end
        end
    end
    ndegen=ones(Float64,1,nr)
    return(rvec,ndegen,ham_r,no,nr)
end

function import_hr(fname)
    tmp=open(fname*"_hr.dat","r") do fp
        readlines(fp)
    end
    no,nr=int(tmp[2]),int(tmp[3])
    return(no,nr)
end

function import_Hopping()
    tmp=open("Hopping.dat","r") do fp
        readlines(fp)
    end
    tmp1=[split(l) for l in tmp[2:5]]
    axis=[[parse(Float64,l) for l in tp] for tp in tmp1]
    tmp1=split(tmp[5])
    no=parse(Int,tmp1[1])
    nr=parse(Int,tmp1[2])
    nor=parse(Int,tmp1[3])
    tmp1=[tmp[no*no*i][:3] for i in 1:nr]
    rvec=[[parse(Float64,l) for l in tp] for tp in tmp1]
    tmp1=[split(l) for l in tmp[(7+no):end]]
    ham_r=reshape([parse(Float64,tp[9])+1im*parse(Float64,tp[10]) for tp in tmp1],nr,no,no)
    return(rvec,ndegen,ham_r,no,nr,axis)
end
